### PR TITLE
Dt 370 jwk server

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Use the JWK server from NomisOauth2.
Switch to Spring Security 5.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -6,7 +6,7 @@ import java.time.LocalDate
 plugins {
     id 'java'
     id 'groovy'
-    id 'org.springframework.boot' version '2.2.4.RELEASE' // remove pinning of tomcat version to 9.0.30 when next upgrading
+    id 'org.springframework.boot' version '2.2.4.RELEASE'
```
Comment (brightonsbox) [inline build.gradle]: Can this be 2.2.5.RELEASE now?

File: src/main/java/net/syscon/elite/security/AuthenticationFacade.java [hunk 2]
```diff
@@ -48,12 +48,9 @@ public boolean isIdentifiedAuthentication() {
         final var auth = getAuthentication();
         return StringUtils.isNotBlank(MDC.get(PROXY_USER))
                 && Optional.ofNullable(auth).
-                filter(OAuth2Authentication.class::isInstance).
```
Comment (andymarke) [inline src/main/java/net/syscon/elite/security/AuthenticationFacade.java]: why was all this replaced?

File: src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationConverter.java [hunk 3]
```diff
@@ -1,43 +1,38 @@
...
-            return AuthorityUtils.commaSeparatedStringToAuthorityList((String) authorities);
-        }
-        if (authorities instanceof Collection) {
-            return AuthorityUtils.commaSeparatedStringToAuthorityList(StringUtils
-                    .collectionToCommaDelimitedString((Collection<?>) authorities));
+    @SuppressWarnings("ConstantConditions")
+    private Collection<GrantedAuthority> extractAuthorities(final Jwt jwt) {
+        final var authorities = new HashSet<>(this.jwtGrantedAuthoritiesConverter.convert(jwt));
+        if (jwt.getClaims().containsKey("authorities")) {
+            authorities.addAll(((Collection<String>) jwt.getClaims().get("authorities")).stream().map(SimpleGrantedAuthority::new).collect(Collectors.toSet()));
```
Comment (andymarke) [inline src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationConverter.java]: an thing we can do about this nasty cast?

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -17,6 +17,15 @@ spring:
     serialization:
       WRITE_DATES_AS_TIMESTAMPS: false
 
+  security:
```
Comment (andymarke) [inline src/main/resources/application.yml]: I guess this is unrelated to the Spring security move but this is what you started of doing?

**Comment by mikehalmamoj** [comment]
> Am I right in thinking if I was running this locally I would also need an Auth running locally as well?
Yes, to verify the signed JWT it gets the public key from Oauth.

File: src/main/java/net/syscon/elite/security/AuthenticationFacade.java (hunk 2 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/security/AuthenticationFacade.java]: OAuth2Authentication is a class from Spring Security Oauth and isn't available in Spring Security 5.  The only line of interest is the call to `isAuthenticated` and we still do that.

File: src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationConverter.java (hunk 3 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationConverter.java]: I could extend the Spring Jwt class to codify our "authorities" claim as a Collection<String>?  Which essentially just moves the cast...  What do you think?

File: src/main/resources/application.yml (hunk 4 above)
Comment (mikehalmamoj) [reply src/main/resources/application.yml]: Yes - started off getting the signing key from the jwk-set-uri location (well, Spring does that for you).  But then added issuer and moved to Spring Security 5 after feedback.

**Pull request merged**